tests/zfs-tests/tests/functional/alloc_class/alloc_class_012_pos.ksh

```diff
@@ -24,6 +24,22 @@
 
 verify_runnable "global"
 
+#
+# Verify the file identified by the input <inode> is written on a special vdev
+# According to the pool layout used in this test vdev_id 3 and 4 are special
+#
+function file_in_special_vdev # <dataset> <inode>
+{
+	typeset dataset="$1"
+	typeset inum="$2"
+
+	zdb -ddddd $dataset $inum | awk '/L0 [0-9]+/ {
+		match($3,"([^:]+):.*",dva);
+		if(dva[1] < 3)
```

Comment:
[stretch goal]
I'd like to see this function become more generic and moved to libtest.shlib.
More specifically, pass in the expected vdev name (eg "mirror-0") as an argument. From "mirror-0" we find the instance (3, in the example here) and compare all of the dvas for the block for inclusion.  This will allow us to more easily write tests that check for copies or otherwise look at where a specific set of blocks are allocated.

Reply:
Without a parsable `zpool status` output (which AFAIK is the only `zpool` output showing redundant devices with unique names) it is not easy to map "mirror-0" to a guid/vdevid number. Parsing `zdb` output is not an option either.

I can see this function becoming more generic once we get http://open-zfs.org/wiki/Vdev_Properties though.

Reply:
It is parseable with the `zdb -C` output. Try something like this:
```
zdb -C testpool |  awk '
  $1=="type:" {type=$2}
  $1=="id:" {id=$2} 
  $1=="guid:" {guid=$2}
  $1=="path:" {path=$2}
  $1~/^children/ {printf("type=%s id=%s guid=%s path=%s\n", type, id, guid, path)} 
  END {printf("type=%s id=%s guid=%s path=%s\n", type, id, guid, path)}
'
``` 
Then if you know the device you are seaching for, it becomes easy to get its id or its parent's id. For the parent you also get the type, so relatively simple to get 'type=mirror'  with  'path=/dev/sda' with 'id=1' where id is also the dva id.

For real-world parsing, we try to directly read the nvlist or convert the nvlist to JSON and parse it with a more object-friendly language like go or python (in other words, not awk or shell). `nvlist_print_json()` is your friend for the conversion, but there is more to understanding the resulting JSON than expected.

In retrospect, it would be nice to have a top-level vdev annotation that is more flexible than booleans like `is_log` but that horse already left the barn.

Reply:
Consider the following layout:

```
root@linux:~# zdb -C testpool |  awk '
  $1=="type:" {type=$2}
  $1=="id:" {id=$2} 
  $1=="guid:" {guid=$2}
  $1=="path:" {path=$2}
  $1~/^children/ {printf("type=%s id=%s guid=%s path=%s\n", type, id, guid, path)} 
  END {printf("type=%s id=%s guid=%s path=%s\n", type, id, guid, path)}
'
type='root' id=0 guid=9264953930665662240 path=
type='mirror' id=0 guid=7373751702954239063 path=
type='file' id=0 guid=7368990276599147716 path='/var/tmp/vdev1'
type='file' id=1 guid=4769494959538226270 path='/var/tmp/vdev2'
type='mirror' id=1 guid=4074419556359544368 path='/var/tmp/vdev2'
type='file' id=0 guid=4562404146963926287 path='/var/tmp/vdev3'
type='file' id=1 guid=7285901949148168926 path='/var/tmp/vdev4'
type='file' id=2 guid=13620745756185228281 path='/var/tmp/vdev5'
```

Is "/var/tmp/vdev5" part of the second mirror (3-way mirror) or a top level vdev?